src/application/service/Wishlist.service.ts

```diff
@@ -0,0 +1,95 @@
+import { Inject, Injectable } from '@nestjs/common';
+import { Wishlist, WishlistGroups, WishlistItem } from 'src/domain/wishlist';
+
+import {
+  ForbiddenWishlistAccessException,
+  WishlistNotFoundException,
+} from '../common/error/exception/wishlist.exception';
+import { ProductDbQueryPort } from '../port/in/product/ProductDbQueryPort';
+import { AddWishlistCommand, WishlistUseCase } from '../port/in/wishlist/WishlistUseCase';
+import { WishlistDbCommandPort } from '../port/out/WishlistDbCommandPort';
+import { WishlistDbQueryPort } from '../port/out/WishlistDbQueryPort';
+
+@Injectable()
+export class WishlistService implements WishlistUseCase {
+  constructor(
+    @Inject('WishlistGateway')
+    private readonly wishlistDbQueryPort: WishlistDbQueryPort,
+    @Inject('WishlistGateway')
+    private readonly wishlistDbCommandPort: WishlistDbCommandPort,
+    @Inject('ProductGateway')
+    private readonly productDbQueryPort: ProductDbQueryPort,
+  ) {}
+
+  async getWishlistGroups(userId: number): Promise<WishlistGroups[]> {
+    const wishlists = await this.wishlistDbQueryPort.getAllByUserId(userId);
+
+    const groupedByReceiver = new Map<string, Wishlist[]>();
+    for (const item of wishlists) {
+      if (!groupedByReceiver.has(item.receiverName)) {
+        groupedByReceiver.set(item.receiverName, []);
+      }
+      groupedByReceiver.get(item.receiverName)!.push(item);
+    }
+
+    const result: WishlistGroups[] = [];
+
+    await Promise.all(
+      Array.from(groupedByReceiver.entries()).map(async ([receiverName, items]) => {
+        const productIds = items.map(w => w.productId);
+
+        const products = await this.productDbQueryPort.getWishlistProductsByIds(productIds);
+
+        const imageUrls = products.map(p => p.imageUrl);
+
+        result.push({
+          receiverName,
+          count: items.length,
+          imageUrls,
+        });
+      }),
+    );
+
+    return result;
```

Comment:
조금 더 개선하자면 receiverName만 가져오도록 변경하면 불필요한 추가 로직도 줄일 수 있고, 디비 부하도 줄일 수 있을 것 같습니다

```suggestion
    const wishlists = await this.wishlistDbQueryPort.getAllByUserId(userId);
    const receiverNames = wishlists.map(wishlist => wishlist.receiverName);

    return Promise.all(
      receiverNames.map(async (receiverName) => {
        const products = await this.productDbQueryPort.getWishlistProductsByReceiverName(receiverName);
        const imageUrls = products.map(p => p.imageUrl);

        return { receiverName, count: products.length, imageUrls };
      }),
    );
```

Reply:
해당 방식으로 수정은 했는데 만약 받는사람이 동일한 상품을 등록해놓은 경우 중복그룹이 생길 수도 있습니다. 그리고 receiverName만 상품을 가져오면 receiverName이 동일할 경우 다른 유저의 상품도 가져와서 userId eh 같이 넣어서 확인하는 것으로 수정했습니다.